---
title: "IQ-TREE / ASTRAL RAD species tree"
author: "Devon DeRaad"
date: '2022-07-15'
output: html_document
---

# workflow describing the process of generating individual gene trees when you did your genotype filtering in R
```{r, eval=FALSE}
#read in vcf with filtered, unlinked loci that we want to use for gene tree estimation
vcfR<-read.vcfR("~/Desktop/todi.2022/todi.unlinked.85.vcf")
#check out the metadata
head(vcfR@fix)
#we want to isolate the third column, which contains the name of each locus
head(vcfR@fix[,3])
#get alist of just locus names to use as whitelist for stacks
whitelist<-sub(":.*", "", vcfR@fix[,3])
#make sure each locus is unique
length(unique(whitelist)) == length(whitelist)
#make sure the locus names look right
whitelist[1:5]

#write out whitelist for stacks
write.table(whitelist, file = "~/Desktop/todi.2022/astral.rad.species.tree/1892.whitelist.txt", quote = F, row.names = F, col.names = F)

#generate popmap including only the samples in this filtered vcf file
#assign each individual to species here and have stacks output phylip with ambiguity codes
df<-data.frame(ind=colnames(vcfR@gt)[-1],
               pop=gsub("_.*","",gsub("T_","",colnames(vcfR@gt)[-1])))

#write out popmap for stacks
write.table(df, file = "~/Desktop/todi.2022/astral.single.tip/8spec.map.txt", quote = F, row.names = F, col.names = F, sep = "\t")
###You then want to randomly downsample this popmap to include only 1 sample per species
#This doesn't allow us to use all the information present in the tree, but it is necessary because we need an alignment and
#input gene trees with only 1 tip per species in order to calculate gCFs and sCFs using IQ-TREE 2

#run this in a terminal window in the directory where you ran populations originally (whitelist includes the loci you want to keep and popmap includes the samples you want to keep based on filtering)
#/home/d669d153/work/stacks-2.41/populations -P /home/d669d153/work/todi.subset.2022/fastq -O . -M 8spec.map.txt --whitelist 1892.whitelist.txt --phylip-var-all
```

# Now convert concatenated phylip into a separate fasta alignment for each locus using R commands and a python script run locally
```{r, eval=FALSE}
#read in the phylip file with all loci concatenated
dna<-read.dna(file = "~/Desktop/todi.2022/astral.single.tip/populations.all.phylip")
as.list(dna)

#write.FASTA(x = dna, file="~/Desktop/aph.data/svdquartets/whole.locus.2725.fasta")

#read in the partition file
parts<-read.table("~/Desktop/todi.2022/astral.single.tip/populations.all.partitions.phylip")
parts$parts<-sub("=.*","",parts$V2)
parts$begin<-sub(".*=","",parts$V2)
parts$begin<-sub("-.*","",parts$begin)
parts$end<-sub(".*=","",parts$V2)
parts$end<-sub(".*-","",parts$end)

#write a loop that generates a separate fasta for each partition
for (i in 1:nrow(parts)){
  locus<- dna[,c(parts$begin[i]:parts$end[i])]
  write.FASTA(x = locus, file=paste0("~/Desktop/todi.2022/astral.single.tip/fastas/locus.",i,".fasta"))
  print(i)
}
```

# now convert each fasta alignment to a nexus file
```{bash, eval=FALSE}
#following steps come from: https://github.com/mmatschiner/tutorials/blob/master/ml_species_tree_inference/README.md#astral
#To remove sequences that contain only missing information from all alignments, and at the same time translate all alignments into Nexus format, we can use the Python script convert.py with the following commands: (run in bash)
cd /Users/devder/Desktop/todi.2022/astral.single.tip
mkdir nex
for i in fastas/*.fasta
do
gene_id=`basename ${i%.fasta}`
python3 convert.fasta.to.nex.py ${i} nex/${gene_id}.nex -f nexus -m 0.9
done

#run this to make sure that missing data is correctly recognized as being encoded by 'N' rather than '?'
cd /Users/devder/Desktop/todi.2022/astral.single.tip/nex
for i in *
  do
sed -i '' 's/ missing=?/ missing=N/' $i
done

#then copy the entire 'nex' directory to the cluster for input to build gene trees using IQTREE 2
```

# now run IQ-TREE 2 to generate individual gene trees, using this submit script for the cluster
```{bash, eval=FALSE}
#We can now use IQ-TREE to generate maximum-likelihood gene trees for all alignments. To be able to later use bootstrapping with ASTRAL,
#we will also generate sets of trees from bootstrapped gene alignments in the same IQ-TREE analysis, by using the option -B 
#we do not specify a substition model with the -m option and therefore allow IQ-TREE to automatically select the best-fitting model.
#we will now ensure that the bootstrap trees are actually written to a file, and not just summarized in the output, by specifying the option --wbt.
#To start IQ-TREE in a loop so that it analyzes one gene alignment after the other, use the following commands: (run on the cluster, 2725 gene trees take ~8 hours to generate on my laptop)

#!/bin/sh
#
#SBATCH --job-name=todi.gene.trees               # Job Name
#SBATCH --nodes=1             # 40 nodes
#SBATCH --ntasks-per-node=1               # 1 CPU allocation per Task
#SBATCH --partition=bi            # Name of the Slurm partition used
#SBATCH --chdir=/home/d669d153/work/todi.subset.2022/astral.single.tip/nex    # Set working d$
#SBATCH --mem-per-cpu=1gb            # memory requested
#SBATCH --time=10000

for i in *.nex
do
/home/d669d153/work/iqtree-2.2.0-Linux/bin/iqtree2 -s ${i} -bb 1000 -wbt -nt AUTO
done

```

# check out the results in an interactive terminal window, and combine output files
```{bash, eval=FALSE}
#The IQ-TREE analyses will have generated a number of files for each gene, containing run info, a distance matrix, starting trees, and so on.
#The only output files required for our further analyses are those ending in .treefile (the maximum-likelihood gene trees with branch lengths)
#and .ufboot (the set of bootstrap trees without branch lengths). To clean up the directory and keep only the important files,
#use the following commands: (run in terminal)
rm nex/*.bionj
rm nex/*.ckp.gz
rm nex/*.contree
rm nex/*.iqtree
rm nex/*.log
rm nex/*.mldist
rm nex/*.model.gz
rm nex/*.splits.nex
rm nex/*.uniqueseq.phy

#Next, have a look at one of the files containing the maximum-likelihood trees, e.g. with the following command:
less nex/locus.1.nex.treefile

#Since ASTRAL will require as input a single file containing all gene trees, combine all files with maximum-likelihood trees into a single file named ml_best.trees, using the following command:
cat nex/*.treefile > ml_best.trees
#To further clean up the directory, you could then also remove all files that contain the maximum-likelihood trees for single genes, using
#rm nex/*.treefile

#We'll first use the set of bootstrapped trees to estimate node support on the species tree.
#To do so, ASTRAL requires as input a single file with the names of all the files containing bootstrapped trees. We can generate such a file with the following command:
ls nex/*.ufboot > ml_boot.txt
```

# Finally, run ASTRAL with and without low confidence branches contracted, as a submit script
```{bash, eval=FALSE}
#!/bin/sh
#
#SBATCH --job-name=todi.gene.trees               # Job Name
#SBATCH --nodes=1             # 40 nodes
#SBATCH --ntasks-per-node=1               # 1 CPU allocation per Task
#SBATCH --partition=bi            # Name of the Slurm partition used
#SBATCH --chdir=/home/d669d153/work/todi.subset.2022/astral.rad.species.tree    # Set working d$
#SBATCH --mem-per-cpu=20gb            # memory requested
#SBATCH --time=10000

#we need a namemap in order to assign tips to species in astral
#easiest to just do by hand unfortunately as the required format is very un-table-like
#save as namemap.txt

#We can then run ASTRAL with two input files: The file containing the maximum-likelihood trees for each gene (ml_best.trees), and the file containing the names of all files with bootstrapped trees (ml_boot.txt).
#The first of these is to be specified with ASTRAL's option -i, and the second should be given with option -b.
#In addition, we'll use option -o to set the name of the output file to species_boot.trees:
module load java
#java -jar /home/d669d153/work/Astral/astral.5.7.7.jar -i ml_best.trees -b ml_boot.txt --outgroup leucopygiu -o todi.astral.boot.trees

#Have a look at the output file species_boot.trees using a text editor (or again the command less). You'll see that it contains 102 lines. 
#The first 100 of these lines represent species trees in Newick format estimated for each the first 100 bootstrapped trees of each gene.
#On line 101 is a consensus tree for these 100 trees. Finally, the last line contains the species tree estimated from the maximum-likelihood gene trees,
#annotated with node support based on the bootstrap trees. Thus, the last line contains the species tree that we'll use for interpretation.
#However, before visualizing the species tree in FigTree, first conduct the second ASTRAL analysis based on the maximum-likelihood trees alone.
#Do so using the following command:
java -jar /home/d669d153/work/Astral/astral.5.7.7.jar -i ml_best.trees --outgroup leucopygiu -o todi.astral.pp.tre
#The output file named species_pp.tre should contain a single species tree, annotated with posterior probabilities as node support.
  
#Since we are interested only in the last of the trees from file species_boot.trees as well as the tree from file species_pp.tre,
#we'll generate a new file named species.trees that contains both of these two trees using the following commands:
tail -n 1 todi.astral.boot.trees > todi.species.boot.tre
#cat todi.astral.pp.tre >> todi.species.trees

#Try contracting branches with extremely low support using newick utilities to see if this improves resolution
/panfs/pfs.local/work/bi/bin/newick_utils/bin/nw_ed  ml_best.trees 'i & b<=10' o > ml_best_BS10.trees

#re-run astral
java -jar /home/d669d153/work/Astral/astral.5.7.7.jar -i ml_best_BS10.trees -a namemap.txt --outgroup leucopygiu -o todi.astral.pp.BS10.tre
```

# calculate gene concordance factors (gCFs) and site concordance factors (sCFs) for your best astral trees using IQTREE 2
```{bash, eval=FALSE}
#!/bin/sh
#
#SBATCH --job-name=todi.gene.trees               # Job Name
#SBATCH --nodes=1             # 40 nodes
#SBATCH --ntasks-per-node=1               # 1 CPU allocation per Task
#SBATCH --partition=bi            # Name of the Slurm partition used
#SBATCH --chdir=/home/d669d153/work/todi.subset.2022/astral.rad.single.tip  # Set working d$
#SBATCH --mem-per-cpu=10gb            # memory requested
#SBATCH --time=10000

#calculate gene concordance factors on ASTRAL guidetree using IQ-TREE 2
/home/d669d153/work/iqtree-2.2.0-Linux/bin/iqtree2 -t todi.astral.pp.tre --gcf ml_best.trees --prefix gene.concordance.pp
/home/d669d153/work/iqtree-2.2.0-Linux/bin/iqtree2 -t todi.astral.pp.BS10.tre --gcf ml_best_BS10.trees --prefix gene.concordance.BS10

#calculate site concordance factors using a concatenated nexus file of all sites as input
#use ruby script to concatenate nexus
#must download NOT COPY AND PASTE this script from here (https://github.com/mmatschiner/tutorials/blob/master/ml_species_tree_inference/src/concatenate.rb)
ruby concatenate.rb -i nex/*.nex -o concatenated.nex -f nexus

#calculate site concordance factors on each tree
/home/d669d153/work/iqtree-2.2.0-Linux/bin/iqtree2 -t todi.astral.pp.tre --gcf ml_best.trees -s concatenated.nex --scf 100 --prefix gene.site.concordance.pp
/home/d669d153/work/iqtree-2.2.0-Linux/bin/iqtree2 -t todi.astral.pp.BS10.tre --gcf ml_best_BS10.trees -s concatenated.nex --scf 100 --prefix gene.site.concordance.BS10
```

# Visualize your two resulting ASTRAL trees labeled with posterior probabilities + gene and site concordance factors, and determine which one you want to use for publication
```{r}
#tree without contracted branches
knitr::include_graphics(c("/Users/devder/Desktop/todi.2022/astral.single.tip/gene.site.concordance.pp.cf.tree.png"))

#tree with contracted branches
knitr::include_graphics(c("/Users/devder/Desktop/todi.2022/astral.single.tip/gene.site.concordance.BS10.cf.tree.png"))
```

# It looks like contracting branches artificially inflates our confidence in resolving deeper branches and biases our estimates of gCFs and sCFs. For this reason, we will use the original tree without contracting low support branches.


# Calculate gCFs and sCFs for nuclear genes on the mitochondrial gene tree and vice versa
```{bash, eval=FALSE}
#calculate gCFs and sCFs based on nuclear gene trees for the mitochondrial species tree
#copy the mito tree into this directory (named mt.species.tre) and make sure the species names match: /home/d669d153/work/todi.subset.2022/astral.rad.single.tip
#then run this
/home/d669d153/work/iqtree-2.2.0-Linux/bin/iqtree2 -t mt.species.tre --gcf ml_best.trees -s concatenated.nex --scf 100 --prefix nuclear.cfs.mito.tree.pp

#calculate gCFs and sCFs based on mitochondrial gene trees for the nuclear species tree
#copy the nuclear tree into this directory (named nuc.species.tre) and make sure the species names match: /home/d669d153/work/todi.subset.2022/astral.rad.single.tip

/home/d669d153/work/iqtree-2.2.0-Linux/bin/iqtree2 -t todi.astral.pp.tre --gcf loci.treefile -s todi.mito.alignment.nex --scf 100 --prefix mito.cfs.nuclear.tree.pp
```

