---
title: "Untitled"
author: "Devon DeRaad"
date: '2023-03-07'
output: html_document
---

```{r, results='hide'}
library(vcfR)
library(SNPfiltR)
library(ggplot2)
library(RColorBrewer)
```

### Filter input SNP files based on smaple inclusion and MAC to increase signal to noise ratio
```{r}
#read in filtered vcf
vcfR <- read.vcfR("~/Desktop/todi.2022/todi.unlinked.85.vcf")
vcfR

#bring in sample info
#read in sample info csv
sample.info<-read.csv("~/Desktop/todi.2022/todiramphus.subset.csv")
sample.info<-sample.info[sample.info$id %in% colnames(vcfR@gt),]
table(sample.info$species)

#see sample order
colnames(vcfR@gt)

#try just removing leucopygius
vcf.2<-vcfR[,c(1:21,27:84)]
colnames(vcf.2@gt)
#remove invariant snps 
vcf.2<-min_mac(vcf.2, min.mac = 1)
#vcfR::write.vcf(vcf.2, file="~/Downloads/todi.noleuco.vcf.gz")
vcf.2

#try removing sanctus and removing leucopygius
vcf.2a<-vcfR[,c(1:21,57:77,83,84)]
colnames(vcf.2a@gt)
#remove invariant snps 
vcf.2a<-min_mac(vcf.2a, min.mac = 1)
#vcfR::write.vcf(vcf.2a, file="~/Downloads/todi.nosanctus.vcf.gz")
vcf.2a

#set up pairwise comparisons between sanctus and other taxa
vcf.2a<-vcfR[,c(1,27:56,78:82,63:70)]
colnames(vcf.2a@gt)
#remove invariant snps 
vcf.2a<-min_mac(vcf.2a, min.mac = 1)
#vcfR::write.vcf(vcf.2a, file="~/Downloads/todi.san.sord.vcf.gz")
vcf.2a

#set up pairwise comparisons between sanctus and other taxa
vcf.2a<-vcfR[,c(1,27:56,78:82,71:77,83,84)]
colnames(vcf.2a@gt)
#remove invariant snps 
vcf.2a<-min_mac(vcf.2a, min.mac = 1)
#vcfR::write.vcf(vcf.2a, file="~/Downloads/todi.san.tris.vcf.gz")
vcf.2a

#set up pairwise comparisons between sanctus and saur
vcf.2a<-vcfR[,c(1,27:56,78:82,57:62)]
colnames(vcf.2a@gt)
#remove invariant snps 
vcf.2a<-min_mac(vcf.2a, min.mac = 1)
#vcfR::write.vcf(vcf.2a, file="~/Downloads/todi.san.saur.vcf.gz")
vcf.2a

```

### Code to convert the vcf into appropriate file structure and run ADMIXTURE on the cluster
```{bash, eval=FALSE}
#!/bin/sh
#
#SBATCH --job-name=admixture.prim              # Job Name
#SBATCH --nodes=1             # 40 nodes
#SBATCH --ntasks-per-node=5               # 40 CPU allocation per Task
#SBATCH --partition=sixhour            # Name of the Slurm partition used
#SBATCH --chdir=/home/d669d153/work/todi.subset.2022/revision.admixture/primary     # Set working d$
#SBATCH --mem-per-cpu=1gb            # memory requested
#SBATCH --time=200

#use plink to convert vcf directly to bed format:
/home/d669d153/work/plink --vcf /home/d669d153/work/todi.subset.2022/revision.admixture/primary/todi.downsamplesanctus.5.vcf  --double-id --allow-extra-chr --make-bed --out binary_fileset
#fix chromosome names
cut -f2- binary_fileset.bim  > temp
awk 'BEGIN{FS=OFS="\t"}{print value 1 OFS $0}' temp > binary_fileset.bim
rm temp

#run admixture for a K of 1-10, using cross-validation, with 10 threads
for K in 1 2 3 4 5 6 7 8 9 10; 
do /home/d669d153/work/admixture_linux-1.3.0/admixture --cv -j5 -m EM binary_fileset.bed $K | tee log${K}.out;
done

#Which K iteration is optimal according to ADMIXTURE ?
grep -h CV log*.out > log.errors.txt
```

### check out the run with all samples included
```{r}
#setwd to admixture directory run on the cluster
setwd("~/Downloads/no.leuco/")
#read in log error values to determine optimal K
log<-read.table("log.errors.txt")[,c(3:4)]
#use double backslash to interpret the opening parentheses literally in the regular expression
log$V3<-gsub("\\(K=", "", log$V3)
log$V3<-gsub("):", "", log$V3)
#interpret K values as numerical
log$V3<-as.numeric(log$V3)
#rename columns
colnames(log)<-c("Kvalue","cross.validation.error")
#make plot showing the cross validation error across K values 1:10
ggplot(data=log, aes(x=Kvalue, y=cross.validation.error, group=1)) +
  geom_line(linetype = "dashed")+
  geom_point()+
  ylab("cross-validation error")+
  xlab("K")+
  scale_x_continuous(breaks = c(1:10))+
  theme_classic()

#read in input file in order to get list of input samples in order
samps<-read.table("binary_fileset.fam")[,1]
#reorder sampling df to match order of the plot
sample.info<-sample.info[match(samps, sample.info$id),]
sample.info$id == samps
#read in all ten runs and save each dataframe in a list
runs<-list()
#read in log files
for (i in 1:10){
  runs[[i]]<-read.table(paste0("binary_fileset.", i, ".Q"))
}
par(mfrow=c(1,1))
#plot each run
for (i in 1:5){
barplot(t(as.matrix(runs[[i]])), col=rainbow(i), ylab="Ancestry", border="black")
}
for (i in 6:10){
barplot(t(as.matrix(runs[[i]])), col=rainbow(i), ylab="Ancestry", border="black")
}

#show sample order
samps
samps[c(1:6,51:56,65:71,77,78,18:20,57:64,7:17,72:76,21:50)]
#color code
brewer.pal(8, "Set2")
barplot(t(as.matrix(runs[[5]]))[,c(1:6,51:56,65:71,77,78,18:20,57:64,7:17,72:76,21:50)], col=c("#E5C494","#A6D854","#FFD92F","#FC8D62","#B3B3B3","#66C2A5","#8DA0CB"), ylab="Ancestry", border="black")
#save barplots
#pdf("~/Desktop/admix.all.pdf", width = 8, height=3)
#barplot(t(as.matrix(runs[[5]]))[,c(1:6,51:56,65:71,77,78,18:20,57:64,7:17,72:76,21:50)], col=c("#E5C494","#A6D854","#FFD92F","#FC8D62","#B3B3B3","#66C2A5","#8DA0CB"), ylab="Ancestry", border="black")
#dev.off()
```

### now check out the results with no sanctus
```{r}
#setwd to admixture directory run on the cluster
setwd("~/Downloads/nosanctus/")
#read in log error values to determine optimal K
log<-read.table("log.errors.txt")[,c(3:4)]
#use double backslash to interpret the opening parentheses literally in the regular expression
log$V3<-gsub("\\(K=", "", log$V3)
log$V3<-gsub("):", "", log$V3)
#interpret K values as numerical
log$V3<-as.numeric(log$V3)
#rename columns
colnames(log)<-c("Kvalue","cross.validation.error")
#make plot showing the cross validation error across K values 1:10
ggplot(data=log, aes(x=Kvalue, y=cross.validation.error, group=1)) +
  geom_line(linetype = "dashed")+
  geom_point()+
  ylab("cross-validation error")+
  xlab("K")+
  scale_x_continuous(breaks = c(1:10))+
  theme_classic()

#read in input file in order to get list of input samples in order
samps<-read.table("binary_fileset.fam")[,1]
#reorder sampling df to match order of the plot
sample.info<-sample.info[match(samps, sample.info$id),]
sample.info$id == samps
#read in all ten runs and save each dataframe in a list
runs<-list()
#read in log files
for (i in 1:10){
  runs[[i]]<-read.table(paste0("binary_fileset.", i, ".Q"))
}
par(mfrow=c(1,1))
#plot each run
for (i in 1:5){
barplot(t(as.matrix(runs[[i]])), col=rainbow(i), ylab="Ancestry", border="black")
}
for (i in 6:10){
barplot(t(as.matrix(runs[[i]])), col=rainbow(i), ylab="Ancestry", border="black")
}

#show sample order
samps
samps[c(1:6,21:26,35:43,18:20,27:34,7,8,12,13,16,17,9:11,14,15)]
#color code
brewer.pal(8, "Set2")
barplot(t(as.matrix(runs[[8]]))[,c(1:6,21:26,35:43,18:20,27:34,7,8,12,13,16,17,9:11,14,15)], col=c("#FFD92F","#66C2A5","black","#8DA0CB","#B3B3B3","#E5C494","#FC8D62","white"), ylab="Ancestry", border="black")

#save barplots
#pdf("~/Desktop/admix.sub.pdf", width = 8, height=3)
#barplot(t(as.matrix(runs[[8]]))[,c(1:6,21:26,35:43,18:20,27:34,7,8,12,13,16,17,9:11,14,15)], #col=c("#FFD92F","#66C2A5","black","#8DA0CB","#B3B3B3","#E5C494","#FC8D62","white"), ylab="Ancestry", border="black")
#dev.off()
```

### now check out pairwise comparisons between sympatric taxa and sanctus
```{r}
#setwd to admixture directory run on the cluster
setwd("~/Downloads/san.sord")
#read in log error values to determine optimal K
log<-read.table("log.errors.txt")[,c(3:4)]
#use double backslash to interpret the opening parentheses literally in the regular expression
log$V3<-gsub("\\(K=", "", log$V3)
log$V3<-gsub("):", "", log$V3)
#interpret K values as numerical
log$V3<-as.numeric(log$V3)
#rename columns
colnames(log)<-c("Kvalue","cross.validation.error")
#make plot showing the cross validation error across K values 1:10
ggplot(data=log, aes(x=Kvalue, y=cross.validation.error, group=1)) +
  geom_line(linetype = "dashed")+
  geom_point()+
  ylab("cross-validation error")+
  xlab("K")+
  scale_x_continuous(breaks = c(1:10))+
  theme_classic()

#read in input file in order to get list of input samples in order
samps<-read.table("binary_fileset.fam")[,1]
#reorder sampling df to match order of the plot
sample.info<-sample.info[match(samps, sample.info$id),]
sample.info$id == samps
#read in all ten runs and save each dataframe in a list
runs<-list()
#read in log files
for (i in 1:10){
  runs[[i]]<-read.table(paste0("binary_fileset.", i, ".Q"))
}
par(mfrow=c(1,1))
#plot each run
for (i in 1:5){
barplot(t(as.matrix(runs[[i]])), col=rainbow(i), ylab="Ancestry", border="black")
}
for (i in 6:10){
barplot(t(as.matrix(runs[[i]])), col=rainbow(i), ylab="Ancestry", border="black")
}

#show sample order
samps
#color code
brewer.pal(8, "Set2")
barplot(t(as.matrix(runs[[2]])), col=c("#A6D854","#E5C494"), ylab="Ancestry", border="black")

#save barplots
#pdf("~/Desktop/admix.san.sord.pdf", width = 8, height=2.8)
#barplot(t(as.matrix(runs[[2]])), col=c("#A6D854","#E5C494"), ylab="Ancestry", border="black")
#dev.off()
```

### now check out pairwise comparisons between sympatric taxa and sanctus
```{r}
#setwd to admixture directory run on the cluster
setwd("~/Downloads/san.tris")
#read in log error values to determine optimal K
log<-read.table("log.errors.txt")[,c(3:4)]
#use double backslash to interpret the opening parentheses literally in the regular expression
log$V3<-gsub("\\(K=", "", log$V3)
log$V3<-gsub("):", "", log$V3)
#interpret K values as numerical
log$V3<-as.numeric(log$V3)
#rename columns
colnames(log)<-c("Kvalue","cross.validation.error")
#make plot showing the cross validation error across K values 1:10
ggplot(data=log, aes(x=Kvalue, y=cross.validation.error, group=1)) +
  geom_line(linetype = "dashed")+
  geom_point()+
  ylab("cross-validation error")+
  xlab("K")+
  scale_x_continuous(breaks = c(1:10))+
  theme_classic()

#read in input file in order to get list of input samples in order
samps<-read.table("binary_fileset.fam")[,1]
#reorder sampling df to match order of the plot
sample.info<-sample.info[match(samps, sample.info$id),]
sample.info$id == samps
#read in all ten runs and save each dataframe in a list
runs<-list()
#read in log files
for (i in 1:10){
  runs[[i]]<-read.table(paste0("binary_fileset.", i, ".Q"))
}
par(mfrow=c(1,1))
#plot each run
for (i in 1:5){
barplot(t(as.matrix(runs[[i]])), col=rainbow(i), ylab="Ancestry", border="black")
}
for (i in 6:10){
barplot(t(as.matrix(runs[[i]])), col=rainbow(i), ylab="Ancestry", border="black")
}

#show sample order
samps
brewer.pal(8, "Set2")
barplot(t(as.matrix(runs[[2]])), col=c("#B3B3B3","#A6D854"), ylab="Ancestry", border="black")

#save barplots
#pdf("~/Desktop/admix.san.tris.pdf", width = 8, height=2.8)
#barplot(t(as.matrix(runs[[2]])), col=c("#B3B3B3","#A6D854"), ylab="Ancestry", border="black")
#dev.off()
```

### now check out pairwise comparisons between sympatric taxa and sanctus
```{r}
#setwd to admixture directory run on the cluster
setwd("~/Downloads/san.saur")
#read in log error values to determine optimal K
log<-read.table("log.errors.txt")[,c(3:4)]
#use double backslash to interpret the opening parentheses literally in the regular expression
log$V3<-gsub("\\(K=", "", log$V3)
log$V3<-gsub("):", "", log$V3)
#interpret K values as numerical
log$V3<-as.numeric(log$V3)
#rename columns
colnames(log)<-c("Kvalue","cross.validation.error")
#make plot showing the cross validation error across K values 1:10
ggplot(data=log, aes(x=Kvalue, y=cross.validation.error, group=1)) +
  geom_line(linetype = "dashed")+
  geom_point()+
  ylab("cross-validation error")+
  xlab("K")+
  scale_x_continuous(breaks = c(1:10))+
  theme_classic()

#read in input file in order to get list of input samples in order
samps<-read.table("binary_fileset.fam")[,1]
#reorder sampling df to match order of the plot
sample.info<-sample.info[match(samps, sample.info$id),]
sample.info$id == samps
#read in all ten runs and save each dataframe in a list
runs<-list()
#read in log files
for (i in 1:10){
  runs[[i]]<-read.table(paste0("binary_fileset.", i, ".Q"))
}
par(mfrow=c(1,1))
#plot each run
for (i in 1:5){
barplot(t(as.matrix(runs[[i]])), col=rainbow(i), ylab="Ancestry", border="black")
}
for (i in 6:10){
barplot(t(as.matrix(runs[[i]])), col=rainbow(i), ylab="Ancestry", border="black")
}

#show sample order
samps
brewer.pal(8, "Set2")
barplot(t(as.matrix(runs[[2]])), col=c("#FFD92F","#A6D854"), ylab="Ancestry", border="black")

#save barplots
#pdf("~/Desktop/admix.san.saur.pdf", width = 8, height=2.8)
#barplot(t(as.matrix(runs[[2]])), col=c("#FFD92F","#A6D854"), ylab="Ancestry", border="black")
#dev.off()
```
